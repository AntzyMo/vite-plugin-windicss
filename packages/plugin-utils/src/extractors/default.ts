import type { ExtractorResultDetailed } from 'windicss/types/interfaces'
import { regexQuotedString, regexClassSplitter, validClassName, regexHtmlTag, regexAttributifyItem } from '../regexes'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function DefaultExtractor(code: string, id?: string): ExtractorResultDetailed {
  if (id?.endsWith('.css') || id?.endsWith('.postcss')) {
    return {
      classes: [],
      tags: [],
    }
  }

  const tags = Array.from(code.matchAll(regexHtmlTag))
  const tagNames = tags.map(i => i[1])

  return {
    tags: tagNames,
    get classes() {
      return Array.from(code.matchAll(regexQuotedString))
        .flatMap(m => (m[2] || '').split(regexClassSplitter))
        .filter(validClassName)
    },
    get attributes() {
      const attributes: ExtractorResultDetailed['attributes'] = {
        names: [],
        values: [],
      }

      tags.forEach((i) => {
        return Array.from(i[2].matchAll(regexAttributifyItem) || [])
          .forEach(([, name,, value]) => {
            attributes.names.push(name)
            attributes.values.push(value)
          })
      })

      return attributes
    },
  }
}
