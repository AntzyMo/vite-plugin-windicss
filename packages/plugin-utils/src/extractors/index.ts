export * from './default'
export * from './helper'
export * from './pug'
export * from './svelte'
