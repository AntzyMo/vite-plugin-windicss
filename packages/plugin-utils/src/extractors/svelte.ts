import type { ExtractorResultDetailed } from 'windicss/types/interfaces'
import { regexSvelteClass } from '../regexes'
import { DefaultExtractor } from './default'

export function SvelteExtractor(code: string, id?: string): ExtractorResultDetailed {
  const result = DefaultExtractor(code, id)

  return {
    tags: result.tags,
    get classes() {
      return [
        ...result.classes!,
        ...Array.from(code.matchAll(regexSvelteClass)).map(i => i[1]).filter(Boolean),
      ]
    },
    get attributes() {
      return result.attributes
    },
  }
}
