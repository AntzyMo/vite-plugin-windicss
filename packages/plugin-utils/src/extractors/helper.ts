import { extname } from 'path'
import type { Extractor } from 'windicss/types/interfaces'
import { DefaultExtractor } from './default'
import { PugExtractor } from './pug'
import { SvelteExtractor } from './svelte'

export function getDefaultExtractors() {
  const extractors: Extractor[] = [
    {
      extractor: SvelteExtractor,
      extensions: ['svelte'],
    },
  ]

  // auto detect pug
  try {
    require('pug')
    extractors.push({
      extractor: PugExtractor,
      extensions: ['vue', 'pug'],
    })
  }
  catch (e) {}

  return extractors
}

export async function applyExtractors(code: string, id?: string, extractors: Extractor[] = [], defaultExtract = DefaultExtractor) {
  let ext = id ? extname(id) : '*'
  if (ext[0] === '.')
    ext = ext.slice(1)
  for (const { extractor, extensions } of extractors) {
    if (extensions.includes(ext))
      return extractor(code, id)
  }
  return defaultExtract(code, id)
}
