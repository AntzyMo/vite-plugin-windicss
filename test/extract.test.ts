import { DefaultExtractor } from '../packages/plugin-utils/src/extractors/default'

describe('extract', () => {
  it('attributify', () => {
    expect(DefaultExtractor(`
      <div
        p="x-4 y-2 md:x-2 lg:hover:x-8"
        sm="bg-red-500 bg-opacity-50 hover:bg-green-300"
        sm:hover="text-red-500 text-lg focus:bg-white"
      />
    `).attributes,
    ).toMatchSnapshot()
  })
})
