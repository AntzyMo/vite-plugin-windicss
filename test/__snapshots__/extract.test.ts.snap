// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract attributify 1`] = `
Object {
  "names": Array [
    "p",
    "sm",
    "sm:hover",
  ],
  "values": Array [
    "x-4 y-2 md:x-2 lg:hover:x-8",
    "bg-red-500 bg-opacity-50 hover:bg-green-300",
    "text-red-500 text-lg focus:bg-white",
  ],
}
`;
